File: crates/docpie/src/org/services.rs
use time::OffsetDateTime;
use uuid::Uuid;

use crate::{
    auth::models::User,
    error::Result,
    http::config::Server,
    org::{
        models::{Org, OrgMembership, OrgMembershipRole, OrgStatus},
        schemas::CreateOrg,
    },
};

pub async fn list_membered_orgs(server: &Server, user: User) -> Result<Vec<Org>> {
    let orgs = sqlx::query_as!(
        Org,
        r#"
        SELECT o.id, o.name, o.slug, o.status AS "status: OrgStatus", o.created_at
        FROM orgs AS o
        JOIN org_membership AS om
            ON om.org_id = o.id
        WHERE om.user_id = $1
        "#,
        user.id,
    )
    .fetch_all(&server.db)
    .await?;

    Ok(orgs)
}

pub async fn create_org(server: &Server, org_to_create: CreateOrg, user: User) -> Result<Org> {
    let mut tx = server.db.begin().await?;

    let org = sqlx::query_as!(
        Org,
        r#"
        INSERT INTO orgs (name, slug, status, created_at)
        VALUES ($1, $2, $3, $4)
        RETURNING id, name, slug, status AS "status: OrgStatus", created_at
        "#,
        org_to_create.name,
        org_to_create.slug,
        OrgStatus::Active as OrgStatus,
        OffsetDateTime::now_utc()
    )
    .fetch_one(&mut *tx)
    .await?;

    let _ = sqlx::query_as!(
        OrgMembership,
        r#"
        INSERT INTO org_membership (user_id, org_id, role, created_at, invite_token)
        VALUES ($1, $2, $3, $4, $5)
        RETURNING id, user_id, org_id, role AS "role: OrgMembershipRole", created_at
        "#,
        user.id,
        org.id,
        OrgMembershipRole::Owner as OrgMembershipRole,
        OffsetDateTime::now_utc(),
        Uuid::now_v7()
    )
    .fetch_one(&mut *tx)
    .await?;

    tx.commit().await?;

    Ok(org)
}

pub async fn get_invite_token(
    server: &Server,
    user: User,
    org_slug: String,
) -> Result<Option<Uuid>> {
    let invite_token = sqlx::query!(
        r#"
        SELECT om.invite_token FROM orgs AS o
        JOIN org_membership AS om
        ON om.org_id = o.id
        WHERE o.slug = $1
        AND om.user_id = $2
        "#,
        org_slug,
        user.id
    )
    .map(|r| r.invite_token)
    .fetch_one(&server.db)
    .await?;

    Ok(invite_token)
}

pub async fn rotate_invite_token(server: &Server, user: User, org_slug: String) -> Result<Uuid> {
    let invite_token = sqlx::query!(
        r#"
        UPDATE org_membership AS om
        SET invite_token = $3
        FROM orgs AS o
        WHERE om.org_id = o.id
        AND o.slug = $1
        AND om.user_id = $2
        RETURNING om.invite_token AS "invite_token!"
        "#,
        org_slug,
        user.id,
        Uuid::now_v7()
    )
    .map(|r| r.invite_token)
    .fetch_one(&server.db)
    .await?;

    Ok(invite_token)
}

File: shared/org/services.ts
import { users } from "#shared/auth/models"
import { db } from "#shared/db"
import type { TOrg, TOrgInsert, TOrgMembership, TOrgMembershipInsert } from "#shared/org/models"
import { orgs, orgMembership } from "#shared/org/models"
import type { TResolvedInvitationOutput } from "#shared/org/schemas"
import { eq, and, inArray } from "drizzle-orm"

export async function listMemberedOrgs(userId: number): Promise<TOrg[]> {
    const memberships = await db.query.orgMembership.findMany({
        where: eq(orgMembership.user_id, userId),
    })

    const orgIds = memberships.map((m) => m.org_id)
    if (orgIds.length === 0) {
        return []
    }

    const result = await db.query.orgs.findMany({
        where: inArray(orgs.id, orgIds),
    })

    return result
}

export async function createOrg(name: string, slug: string, userId: number): Promise<TOrg> {
    const result = await db.transaction(async (tx) => {
        const newOrg: TOrgInsert = {
            name,
            slug,
            status: "active",
        }

        const orgResult = await tx.insert(orgs).values(newOrg).returning()
        const org = orgResult[0]

        if (!org) {
            throw new Error("Failed to create organization")
        }

        const newMembership: TOrgMembershipInsert = {
            user_id: userId,
            org_id: org.id,
            role: "owner",
            invite_token: crypto.randomUUID(),
        }

        await tx.insert(orgMembership).values(newMembership)

        return org
    })

    return result
}

export async function getInviteToken(userId: number, orgSlug: string): Promise<string | null> {
    const org = await db.query.orgs.findFirst({
        where: eq(orgs.slug, orgSlug),
    })

    if (!org) {
        return null
    }

    const membership = await db.query.orgMembership.findFirst({
        where: and(eq(orgMembership.user_id, userId), eq(orgMembership.org_id, org.id)),
    })

    return membership?.invite_token ?? null
}

export async function rotateInviteToken(userId: number, orgSlug: string): Promise<string | null> {
    const org = await db.query.orgs.findFirst({
        where: eq(orgs.slug, orgSlug),
    })

    if (!org) {
        return null
    }

    const newToken = crypto.randomUUID()

    const result = await db
        .update(orgMembership)
        .set({ invite_token: newToken })
        .where(and(eq(orgMembership.user_id, userId), eq(orgMembership.org_id, org.id)))
        .returning()

    if (result.length === 0) {
        return null
    }

    return newToken
}

export async function resolveInviteToken(
    orgSlug: string,
    token: string,
): Promise<TResolvedInvitationOutput | null> {
    const org = await db.query.orgs.findFirst({
        where: eq(orgs.slug, orgSlug),
    })

    if (!org) {
        return null
    }

    const membership = await db.query.orgMembership.findFirst({
        where: and(eq(orgMembership.org_id, org.id), eq(orgMembership.invite_token, token)),
    })

    if (!membership) {
        return null
    }

    const inviter = await db.query.users.findFirst({
        where: eq(users.id, membership.user_id),
    })

    if (!inviter) {
        return null
    }

    return {
        org_name: org.name,
        inviter_name: inviter.display_name,
        invite_token: token,
    }
}

export async function isOrgSlugTaken(slug: string): Promise<boolean> {
    const org = await db.query.orgs.findFirst({
        where: eq(orgs.slug, slug),
    })
    return org !== null
}

export async function getUserMembership(
    userId: number,
    orgId: number,
): Promise<TOrgMembership | null> {
    const membership = await db.query.orgMembership.findFirst({
        where: and(eq(orgMembership.user_id, userId), eq(orgMembership.org_id, orgId)),
    })
    return membership ?? null
}
